src/main/java/com/hubspot/smtp/client/SmtpSessionFactory.java

```diff
@@ -42,16 +45,26 @@ public SmtpSessionFactory(ExecutorService executorService) {
     bootstrap.connect().addListener(f -> {
       if (f.isSuccess()) {
         SmtpSession session = new SmtpSession(((ChannelFuture) f).channel(), responseHandler, executorService);
-        initialResponseFuture.thenAccept(r -> connectFuture.complete(new SmtpClientResponse(r[0], session)));
+        applyOnExecutor(initialResponseFuture, r -> connectFuture.complete(new SmtpClientResponse(r[0], session)));
       } else {
         LOG.error("Could not connect to {}", config.getRemoteAddress(), f.cause());
-        connectFuture.completeExceptionally(f.cause());
+        executorService.execute(() -> connectFuture.completeExceptionally(f.cause()));
       }
     });
 
     return connectFuture;
   }
 
+  private <R, T> CompletableFuture<R> applyOnExecutor(CompletableFuture<T> eventLoopFuture, Function<T, R> mapper) {
+    return eventLoopFuture.handleAsync((rs, e) -> {
+      if (e != null) {
```

Comment:
1. probably worth noting here that the reason why you're using handleAsync and propagating the error rather than using .thenApplyAsync directly is to get the completion running on another thread in an exceptional case, right? If so I would probably add a one-line comment

2. I think we should move towards using `throw new RuntimeException()` everywhere